components/cdc/src/old_value.rs

```diff
@@ -0,0 +1,377 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::ops::Deref;
+
+use engine_traits::{ReadOptions, CF_DEFAULT, CF_WRITE};
+use tikv::storage::{Cursor, CursorBuilder, ScanMode, Snapshot as EngineSnapshot, Statistics};
+use tikv_util::config::OptionReadableSize;
+use tikv_util::lru::{CacheSizeTrace, RawLruCache};
+use tikv_util::sys::sys_quota::SysQuota;
+use tikv_util::time::Instant;
+use txn_types::{Key, MutationType, OldValue, TimeStamp, Value, WriteRef, WriteType};
+
+use crate::metrics::*;
+use crate::Result;
+
+pub fn get_old_value<S: EngineSnapshot>(
+    key: Key,
+    query_ts: TimeStamp,
+    old_value_cache: &mut OldValueCache,
+    reader: &OldValueReader<S>,
+) -> (Option<Value>, Option<Statistics>) {
+    if let Some((old_value, mutation_type)) = old_value_cache.cache.remove(&key) {
+        return match mutation_type {
+            MutationType::Insert => (None, None),
+            MutationType::Put | MutationType::Delete => match old_value {
+                OldValue::None => (None, None),
+                OldValue::Value {
+                    start_ts,
+                    short_value,
+                } => {
+                    let mut statistics = None;
+                    let value = short_value.or_else(|| {
+                        statistics = Some(Statistics::default());
+                        let prev_key = key.truncate_ts().unwrap().append_ts(start_ts);
+                        let start = Instant::now();
+                        let mut opts = ReadOptions::new();
+                        opts.set_fill_cache(false);
+                        let value =
+                            reader.get_value_default(&prev_key, statistics.as_mut().unwrap());
+                        CDC_OLD_VALUE_DURATION_HISTOGRAM
+                            .with_label_values(&["get"])
+                            .observe(start.elapsed().as_secs_f64());
+                        value
+                    });
+                    (value, statistics)
+                }
+                // Unspecified should not be added into cache.
+                OldValue::Unspecified => unreachable!(),
+            },
+            _ => unreachable!(),
+        };
+    }
+    // Cannot get old value from cache, seek for it in engine.
+    let mut statistics = Statistics::default();
+    let start = Instant::now();
+    let key = key.truncate_ts().unwrap().append_ts(query_ts);
+    let value = reader
+        .near_seek_old_value(&key, &mut statistics)
+        .unwrap_or_default();
+    CDC_OLD_VALUE_DURATION_HISTOGRAM
+        .with_label_values(&["seek"])
+        .observe(start.elapsed().as_secs_f64());
+    (value, Some(statistics))
+}
+
+pub(crate) type OldValueCallback =
+    Box<dyn Fn(Key, TimeStamp, &mut OldValueCache) -> (Option<Vec<u8>>, Option<Statistics>) + Send>;
+
+#[derive(Default)]
+struct OldValueCacheSizeTrace(u64);
```

Comment:
It should be a signed integer, otherwise, we may observe an increasing curve in dashboards.

Reply:
I think expectedly it should never be a negative number.